### PR TITLE
Updates

### PR DESCRIPTION
**Pull request by Sarahtonein**
Added the following functionality:

``` !listevents ``` 
 command list ALL events past and present that are in the Bloom Discord server

``` !deleteevent ``` 
command deletes an event. It needs to be passed an event ID with the command.

Scheduled task that will check for events within the next 24 hours once a day. 
If events are found it will log them to a specified channel and use the @ here to inform people of the upcoming event/s

When new events are created, there is an hour delay in which the person creating the event can make changes to the name / description etc.

### CODE REVIEW

**Comment by Sarahtonein** [comment]
To clarify on !listevents, it will only list events that are still visible within the Bloom Discord server. Once an event has been ended / deleted it will no longer show.

**Comment by Sarahtonein** [comment]
@Balugrizzly , same with this branch

**Comment by Balugrizzly** [comment]
same as what with this branch?

**Comment by Balugrizzly** [comment]
You have linked the help ticket to this which will be closed after merge is this intended?
The help command should exist for the whole bot in its entirety

**Comment by Balugrizzly** [comment]
merge conflicts

File: CHANGELOG.md
Comment (Balugrizzly) [inline CHANGELOG.md:1]: I assume you have added v0.1 as a placeholder? and are aware that there is no v01

File: Pipfile
Comment (Balugrizzly) [inline Pipfile:9]: Why are you installing dotenv?

File: main.py
Comment (Balugrizzly) [inline main.py:1]: Can you separate the main entry point from the actual functionality and let it branch out

**Review by Balugrizzly — CHANGES_REQUESTED** [review]
Consider opening your files once instead of multiple times and duplicating the functionality.
Same goes for similar operations

File: updates/updates.py
Comment (Sarahtonein) [reply updates/updates.py:53]: > Why are you waiting 30 min here?

Arbitrary wait time to allow for editing of the event prior to posting.
This could be greatly shortened.

File: updates/updates.py
Comment (Sarahtonein) [reply updates/updates.py]: > This comment makes no sense

Have rephrased the comment

**Comment by Sarahtonein** [comment]
> Consider opening your files once instead of multiple times and duplicating the functionality. Same goes for similar operations

Ah I see what you mean. 👍🏼

File: CHANGELOG.md
Comment (Sarahtonein) [reply CHANGELOG.md:1]: > I assume you have added v0.1 as a placeholder? and are aware that there is no v01

Yes just a placeholder

File: Pipfile
Comment (Sarahtonein) [reply Pipfile:9]: > Why are you installing dotenv?

I was using it initially but it can be removed I think

File: updates/updates.py
Comment (Sarahtonein) [reply updates/updates.py:74]: > Not sure what you are doing with this function and here

Have removed this and am using guild.scheduled_events instead

File: main.py
Comment (Balugrizzly) [reply main.py:1]: bump

File: updates/updates.py
Comment (Balugrizzly) [reply updates/updates.py:50]: Is that the hour delay that was mentioned it discord?
Avoid a situation where you have two 30min waits

File: updates/updates.py
Comment (Balugrizzly) [inline updates/updates.py:68]: What is the purpose of this function?
Whats the explanation for time_range and it being set to None

File: updates/updates.py
Comment (Balugrizzly) [inline updates/updates.py:84]: Pls explain this section 
You are appending your event in any case

File: updates/updates.py
Comment (Balugrizzly) [inline updates/updates.py:86]: As said in the previous original comment 
This does not make sense to me

File: README.md
Comment (Balugrizzly) [inline README.md:9]: This does not match or seemingly not match witht he only 30mins delay i found and commented on

File: README.md
Comment (Balugrizzly) [inline README.md:62]: Is this still accurate?
if so how are people getting the event id?
if they have to go into the events well ... :)

**Review by Balugrizzly — CHANGES_REQUESTED** [review]

**Comment by Sarahtonein** [comment]
> same as what with this branch?

Oh I was just mentioning that this PR was updated from the putting_it_together branch. You can disregard that comment.

File: README.md
Comment (Sarahtonein) [reply README.md:62]: > Is this still accurate? if so how are people getting the event id? if they have to go into the events well ... :)

There is a bot command ``list_events`` which provides this detail. 

``` 
@bot.command(name='list_events')
async def listevents(ctx):
    guild = ctx.guild
    event_list = await check_upcoming_events(guild)

    # Extracting event information
    event_info_list = [(event.name, event.id, event.description) for event in event_list]

    # Formatting the information
    formatted_events = [
        f"🌺 **{name}**🌺 \n**event_id: **{event_id}\n**Description:** {description}"
        for name, event_id, description in event_info_list
    ]
    formatted_string = "\n\n".join(formatted_events)

    await ctx.send(f"🗓️ **All Events**🗓️ \n\n{formatted_string}")

```

File: README.md
Comment (Sarahtonein) [reply README.md:9]: > This does not match or seemingly not match witht he only 30mins delay i found and commented on

Yeah, I originally had this as an hour. But then I also had the check for upcoming events within the next 24 hours set to the top of an hour so I changed to 30 minutes so that there wasn't two things running around the same time.

I will update this to reflect that change

File: main.py
Comment (Sarahtonein) [reply main.py:1]: > Can you separate the main entry point from the actual functionality and let it branch out

Yup will do so!

File: updates/updates.py
Comment (Sarahtonein) [reply updates/updates.py:50]: > Is that the hour delay that was mentioned it discord? Avoid a situation where you have two 30min waits

Updated the Readme, originally I had set the delay to an hour but then it was suggested to change the upcoming events to be at the top of the hour, so this was changed to 30 minutes.

**Comment by Sarahtonein** [comment]
> You have linked the help ticket to this which will be closed after merge is this intended? The help command should exist for the whole bot in its entirety

I intended to handle that issue in this branch / PR, but I can create another branch for that issue specifically?

File: updates/updates.py
Comment (Sarahtonein) [reply updates/updates.py:68]: > What is the purpose of this function? Whats the explanation for time_range and it being set to None

With time_range set to none by default it was intended to capture all events regardless of what time they start in the future by default, and then performing a time comparison to see if they start within the next 24 hours.

This function can be removed I think and some of the logic implemented into the task itself.

File: updates/updates.py
Comment (Sarahtonein) [reply updates/updates.py:84]: > Pls explain this section You are appending your event in any case

I was thinking if we were going to implement something else related to the events we may want to return them anyway. But I will rework this logic per my other comment about the purpose / logic of the function

File: updates/updates.py
Comment (Sarahtonein) [reply updates/updates.py]: This function has been removed entirely. 

> As said in the previous original comment This does not make sense to me

**Comment by Sarahtonein** [comment]
> Consider opening your files once instead of multiple times and duplicating the functionality. Same goes for similar operations

I was thinking about this and as its a small file wouldn't it be better to open and close as it's just a small file? Wouldn't there potentially be data loss if the bot was restarted / crashed and the file had not been closed / saved?

**Comment by Balugrizzly** [comment]
> > You have linked the help ticket to this which will be closed after merge is this intended? The help command should exist for the whole bot in its entirety
> 
> I intended to handle that issue in this branch / PR, but I can create another branch for that issue specifically?

As long as the help command is available for everything its ok

**Comment by Balugrizzly** [comment]
> > Consider opening your files once instead of multiple times and duplicating the functionality. Same goes for similar operations
> 
> I was thinking about this and as its a small file wouldn't it be better to open and close as it's just a small file? Wouldn't there potentially be data loss if the bot was restarted / crashed and the file had not been closed / saved?

What you state  is unrelated to what i mentioned 
The size of the file has no significance in this.
May i suggest implementing it in a way that data loss by design does not occur :) I think that would be beneficial
The implementation is not the only thing the duplication and the referencing or the lack there of is one

**Review by Balugrizzly — COMMENTED** [review]
Can you explain your directory structure layout
If you want we can postpone this and revisit when you bring the three branches together
I would lean towards a domain separation or responsibility sections

**Comment by Sarahtonein** [comment]
> Can you explain your directory structure layout If you want we can postpone this and revisit when you bring the three branches together I would lean towards a domain separation or responsibility sections

I was thinking of separating into:

Discord package related:

Commands / Events / Tasks

And then separate folders for:

Emojis/Emotes

Updates/Events

Governance/Proposals

Snapshot

File: constants.py
Comment (Balugrizzly) [inline constants.py:1]: Why are you using config if you are making them into constants?
Why not use constants

File: main.py
Comment (Balugrizzly) [inline main.py:24]: ? U importing main anywhere :)

Comment (Sarahtonein) [reply main.py:24]: > ? U importing main anywhere :)

Not in this branch and I don't think in others.

File: constants.py
Comment (Sarahtonein) [reply constants.py:1]: > Why are you using config if you are making them into constants? Why not use constants

I thought it would look cleaner to have the values stored in the config instead but I see your point

**Review by Balugrizzly — COMMENTED** [review]
Up to you on how you want to proceed

**Comment by Sarahtonein** [comment]
> Up to you on how you want to proceed

Happy to merge both of updates and emotes after my changes yesterday. Tested both and are OK.

**Pull request merged**
